pyrefly/lib/alt/call.rs

```diff
@@ -1101,4 +1133,186 @@ impl<'a, Ans: LookupAnswer> AnswersSolver<'a, Ans> {
             self.unions(vec![new_attr_ty, init_attr_ty])
         }
     }
+
+    /// Promote the return type of certain string methods to LiteralString when all arguments
+    /// are LiteralString, per PEP 675. This handles format(), join(), and replace().
+    fn promote_literal_string_method_return(
+        &self,
+        method_name: &Name,
+        self_obj: &Type,
+        args: &[CallArg],
+        keywords: &[CallKeyword],
+        return_ty: Type,
+    ) -> Type {
+        // Check for starred args or **kwargs first - these prevent promotion
+        // even if the overload selected returns LiteralString
+        for arg in args {
+            if matches!(arg, CallArg::Star(..)) {
+                // Starred args mean we can't statically verify all args are literal strings
+                return if return_ty.is_literal_string() {
+                    // Demote from LiteralString to str
+                    Type::ClassType(self.stdlib.str().clone())
+                } else {
+                    return_ty
+                };
+            }
+        }
+        for kw in keywords {
+            if kw.arg.is_none() {
+                // **kwargs unpack
+                return if return_ty.is_literal_string() {
+                    // Demote from LiteralString to str
+                    Type::ClassType(self.stdlib.str().clone())
+                } else {
+                    return_ty
+                };
+            }
+        }
+
+        // Early returns for cases where promotion doesn't apply
+        if return_ty.is_literal_string() {
+            return return_ty;
+        }
+        if !self_obj.is_literal_string() {
+            return return_ty;
+        }
+        if !matches!(return_ty, Type::ClassType(ref cls) if cls.is_builtin("str")) {
+            return return_ty;
+        }
+
+        // Error collector for re-inferring (never extended to main errors)
+        let error_swallower = self.error_collector();
+
+        match method_name.as_str() {
+            "format" => {
```

Comment:
My hunch is that maybe we don’t need to be super specific here with the argument count, and we can just have a single block handle format/replace that checks all the args/kwargs are LiteralString.

Since LiteralString is special-cased in the typechecker, I don’t think there's a risk of conflating these methods with different methods w/ the same name.

It might be good to put the starred arg/kwarg checks down here too, jsut the arg/kwarg checks are grouped together.